File: app/Models/CreditTransferRequest.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class CreditTransferRequest extends Model
{
    /**
     * Fixed hour value per leadership position, per ข้อ 14 of the
     * university announcement — students don't choose the hours, they
     * choose the position and the hours follow.
     */
    public const POSITION_HOURS = [
        'student_council_president' => 60,
        'student_club_president' => 60,
        'student_parliament_president' => 60,
        'club_president' => 50,
        'dormitory_president' => 50,
        'class_leader' => 50,
        'class_representative' => 50,
    ];

    /**
     * Thai labels for the keys above — single source, reused by both admin
     * (resources/views/admin/credit-transfers/index.blade.php,
     * admin/students/show.blade.php) and student-facing dashboards
     * (Student\DashboardController, Api\Student\DashboardController), which
     * previously each hand-maintained their own identical copy.
     */
    public const POSITION_LABELS = [
        'student_council_president' => 'นายกองค์การบริหารนักศึกษา',
        'student_club_president' => 'นายกสโมสรนักศึกษา',
        'student_parliament_president' => 'ประธานสภานักศึกษา',
        'club_president' => 'ประธานชมรม',
        'dormitory_president' => 'ประธานหอพักมหาวิทยาลัย',
        'class_leader' => 'หัวหน้าหมู่เรียน',
        'class_representative' => 'ตัวแทนหมู่เรียน',
    ];

    protected $fillable = [
        'user_id',
        'position',
        'academic_year',
        'hours_requested',
        'hours_approved',
        'activity_category',
        'proof_image_path',
        'status',
        'reject_reason',
        'admin_comment',
        'reviewed_by',
        'reviewed_at',
    ];

    protected function casts(): array
    {
        return [
            'reviewed_at' => 'datetime',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function reviewer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reviewed_by');
    }

    public function scopePending($query)
    {
        return $query->where('status', 'pending');
    }

    /**
     * The hours actually credited: what an admin overrode to, or the
     * position's standard hours if never adjusted.
     */
    public function getHoursCreditedAttribute(): int
    {
        return $this->hours_approved ?? $this->hours_requested;
    }

    /**
     * Whether a student already has a pending or approved credit-transfer
     * request for the given academic year — the announcement allows only
     * one claim per academic year. A rejection frees the slot back up.
     */
    public static function hasClaimedAcademicYear(int $userId, int $academicYear): bool
    {
        return static::where('user_id', $userId)
            ->where('academic_year', $academicYear)
            ->whereIn('status', ['pending', 'approved'])
            ->exists();
    }
}

File: mobile/lib/core/models/credit_transfer_request.dart
class CreditTransferPosition {
  CreditTransferPosition({required this.key, required this.label, required this.hours});

  final String key;
  final String label;
  final int hours;

  factory CreditTransferPosition.fromJson(Map<String, dynamic> json) {
    return CreditTransferPosition(
      key: json['key'] as String,
      label: json['label'] as String,
      hours: json['hours'] as int,
    );
  }
}

class CreditTransferRequest {
  CreditTransferRequest({
    required this.id,
    required this.position,
    required this.positionLabel,
    required this.academicYear,
    required this.hoursRequested,
    this.hoursApproved,
    required this.hoursCredited,
    required this.status,
    this.rejectReason,
  });

  final int id;
  final String position;
  final String positionLabel;
  final int academicYear;
  final int hoursRequested;
  final int? hoursApproved;
  final int hoursCredited;
  final String status;
  final String? rejectReason;

  factory CreditTransferRequest.fromJson(Map<String, dynamic> json) {
    return CreditTransferRequest(
      id: json['id'] as int,
      position: json['position'] as String,
      positionLabel: json['position_label'] as String,
      academicYear: json['academic_year'] as int,
      hoursRequested: json['hours_requested'] as int,
      hoursApproved: json['hours_approved'] as int?,
      hoursCredited: json['hours_credited'] as int,
      status: json['status'] as String,
      rejectReason: json['reject_reason'] as String?,
    );
  }
}
